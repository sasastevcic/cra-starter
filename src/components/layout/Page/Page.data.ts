import { ComponentPropsWithoutRef, ReactNode } from 'react';

export interface PageProps extends ComponentPropsWithoutRef<'div'> {
	children: ReactNode;
}
