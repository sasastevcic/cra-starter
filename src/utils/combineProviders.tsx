/* eslint-disable @typescript-eslint/no-explicit-any */
import { ComponentProps, ComponentType, FunctionComponent, ReactNode } from 'react';

type Providers = Array<[ComponentType<any>, ComponentProps<any>?]>;

export const combineProviders = (
	providers: Providers,
): FunctionComponent<{ children: ReactNode }> =>
	providers.reduce(
		(AccumulatedProviders, [Provider, props = {}]) =>
			({ children }) =>
				(
					<AccumulatedProviders>
						<Provider {...props}>{children}</Provider>
					</AccumulatedProviders>
				),
		({ children }) => <>{children}</>,
	);
