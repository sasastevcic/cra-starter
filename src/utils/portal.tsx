import { ReactNode, ReactPortal, useMemo } from 'react';
import { createPortal } from 'react-dom';

interface Portal {
	portalId: string;
	children: ReactNode;
}

const Portal = ({ portalId, children }: Portal): ReactPortal => {
	const portalElement = useMemo(() => document.getElementById(portalId), [portalId]);

	if (!portalElement) {
		throw new Error(`Element not found! Element id: ${portalId}`);
	}

	return createPortal(children, portalElement);
};

export default Portal;
