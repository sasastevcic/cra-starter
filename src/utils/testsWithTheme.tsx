/* eslint-disable @typescript-eslint/ban-ts-comment */
import { ReactElement, ComponentType, ReactNode } from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { theme } from '../styles/config/theme';

export const componentWithTheme = (component: ReactNode): ReactElement => (
	// @ts-ignore
	<ThemeProvider theme={theme}>{component}</ThemeProvider>
);

export const renderWithTheme = (component: ReactElement, options?: RenderOptions): RenderResult => {
	const wrapper: ComponentType<{ children: ReactNode }> = ({ children }) =>
		componentWithTheme(children);

	return render(component, { wrapper, ...options });
};
