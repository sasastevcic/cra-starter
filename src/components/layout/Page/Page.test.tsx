import renderer from 'react-test-renderer';
import { renderWithTheme } from '../../../utils/testsWithTheme';
import Noop from '../../atoms/Noop';
import Page from './index';

describe('Page', () => {
	describe('Render', () => {
		it('should render', () => {
			const { getByTestId } = renderWithTheme(
				<Page>
					<Noop />
				</Page>,
			);

			const element = getByTestId('Page');

			expect(element).toBeInTheDocument();
		});

		it('should render correctly', () => {
			const tree = renderer
				.create(
					<Page>
						<Noop />
					</Page>,
				)
				.toJSON();

			expect(tree).toMatchSnapshot();
		});
	});

	describe('Content', () => {
		it('should have elements as children', () => {
			const { getByTestId } = renderWithTheme(
				<Page>
					<span>Child</span>
				</Page>,
			);

			const element = getByTestId('Page');
			const child = element.querySelector('span');

			expect(element).toContainElement(child);
		});
	});
});
