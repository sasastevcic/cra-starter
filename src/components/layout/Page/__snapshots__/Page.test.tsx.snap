// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page Render should render correctly 1`] = `
<div
  className="sc-bczRLJ ljkdgA"
  data-testid="Page"
/>
`;
